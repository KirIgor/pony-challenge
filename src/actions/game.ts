// import { Direction } from '../types';
// import MockPonyApi from '../api/mock_pony_api';
// import { parseCharactersPosition } from '../utils/parser';
// import { Dispatch } from 'react-redux';
//
// // actions
// export const UPDATE_CHARACTERS_POSITION = 'UPDATE_CHARACTERS_POSITION';
// export type UPDATE_CHARACTERS_POSITION = typeof UPDATE_CHARACTERS_POSITION;
//
// // action creators
// export interface Move {
// 	type: UPDATE_CHARACTERS_POSITION;
//   payload: { }
// }

// export const move = (direction: Direction) => (dispatch, getState) => {
// return MockPonyApi.makeMove(direction).then(apiState => {
//   dispatch({
//     type: UPDATE_CHARACTERS_POSITION
//     payload: {}
//   });
// });
//
// }
