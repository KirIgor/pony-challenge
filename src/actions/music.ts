import { ActionCreator } from 'redux';

// action types
export const TOGGLE_MUSIC = 'TOGGLE_MUSIC';
export type TOGGLE_MUSIC = typeof TOGGLE_MUSIC;

// actions

export interface ToggleMusic {
	type: TOGGLE_MUSIC;
}

export type MusicAction = ToggleMusic;

// action creators

export const toggleMusic: ActionCreator<ToggleMusic> = () => ({ type: TOGGLE_MUSIC });
