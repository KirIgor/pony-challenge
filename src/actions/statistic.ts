import { ActionCreator } from 'redux';

// action types
export const ADD_WINS = 'ADD_WINS';
export type ADD_WINS = typeof ADD_WINS;

export const ADD_LOSES = 'ADD_LOSES';
export type ADD_LOSES = typeof ADD_LOSES;

// actions

export interface AddWins {
	type: ADD_WINS;
}

export interface AddLoses {
	type: ADD_LOSES;
}

export type StatisticAction = AddWins | AddLoses;

// action creators

export const addWins: ActionCreator<AddWins> = () => ({ type: ADD_WINS });
export const addLoses: ActionCreator<AddLoses> = () => ({ type: ADD_LOSES });
