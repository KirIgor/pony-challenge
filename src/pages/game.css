.game-container {
	width: 100%;
	height: -webkit-fill-available;
	display: flex;
	justify-content: center;
	align-items: center;
}
