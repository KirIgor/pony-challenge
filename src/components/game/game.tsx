import * as React from 'react';
import { connect } from 'react-redux';
import { Dispatch, bindActionCreators } from 'redux';

import { PonyName, Direction, RainbowPath, Role } from '../../types/index';
import { init, move } from '../../actions/game';
import { StoreState } from '../../store/store';
import { GameState } from '../../types/index';
import Maze from '../maze/maze';

interface Props {
	gameState: GameState;
	init: typeof init;
	move: typeof move;
}

const onMove = (moveF: typeof move) => (
	gameState: GameState,
	setPath: React.Dispatch<React.SetStateAction<RainbowPath>>
) => (e: React.KeyboardEvent) => {
	const direction = keyCodeToDirection(e.keyCode);
	if (direction) {
		moveF(direction);

		const { x, y } = gameState.charactersPosition.get(Role.PONY)!;
		setPath(prev => prev.slice(-1).concat({ x, y, direction }) as RainbowPath);
	}
};

function Game({ gameState, init, move }: Props) {
	React.useEffect(() => {
		init(15, 15, PonyName.APPLEJACK, 10);
	}, []);

	const [rainbowPath, setPath] = React.useState<RainbowPath>([]);

	return (
		<div
			ref={node => node && node.focus()}
			tabIndex={0}
			onKeyDown={onMove(move)(gameState, setPath)}>
			<Maze gameState={gameState} ponyName={PonyName.APPLEJACK} rainbowPath={rainbowPath} />
		</div>
	);
}

const keyCodeToDirection = (keyCode: number): Direction | undefined => {
	switch (keyCode) {
		case 37:
		case 65: {
			return Direction.WEST;
		}
		case 38:
		case 87: {
			return Direction.NORTH;
		}
		case 39:
		case 68: {
			return Direction.EAST;
		}
		case 40:
		case 83: {
			return Direction.SOUTH;
		}
		default:
			return undefined;
	}
};

const mapStateToProps = (state: StoreState) => ({ gameState: state.game });
const mapDispatchToProps = (dispatch: Dispatch) => ({
	init: bindActionCreators(init, dispatch),
	move: bindActionCreators(move, dispatch)
});

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(Game);
