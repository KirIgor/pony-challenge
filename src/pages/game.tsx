import * as React from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
// import { Link } from 'react-router-dom';
// import * as Routes from '../constants/routes.json';

import { parseQuery } from '../utils/helper';
import { PonyName } from '../types/index';
import { StoreState } from '../store/store';
import Loadable from '../components/loadable/loadable';
import { init, GameAction } from '../actions/game';
import GameComponent from '../components/game/game';

import './game.css';

type Props = ReturnType<typeof mapDispatchToProps> & {
	width: number;
	height: number;
	ponyName: PonyName;
	difficulty: number;
	history: any;
};

const Game = React.memo(({ width, height, ponyName, difficulty, history, init }: Props) => {
	return (
		<div className="game-container">
			<Loadable load={init(width, height, ponyName, difficulty)}>
				<GameComponent ponyName={ponyName} newGameFunction={() => history.goBack(1)} />
			</Loadable>
		</div>
	);
});

const mapStateToProps = (_: any, own: any) => {
	const query = parseQuery(own.location.search) as any;

	return {
		width: parseInt(query.width),
		height: parseInt(query.height),
		ponyName: query.ponyName,
		difficulty: parseInt(query.difficulty)
	};
};

// in lack of redux-thunk pull request #224
const mapDispatchToProps = (dispatch: ThunkDispatch<StoreState, void, GameAction>) => ({
	init: (width: number, height: number, playerName: PonyName, difficulty: number) =>
		dispatch(init(width, height, playerName, difficulty))
});

export default connect(
	mapStateToProps,
	mapDispatchToProps
)(Game);
