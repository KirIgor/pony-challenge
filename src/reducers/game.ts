import { Map as iMap, List as iList } from 'immutable';

import { GameState, BlueprintRecord, Role, Point, GameStatus } from '../types/index';
import { GameAction, UPDATE_STATE, INIT_STATE } from '../actions/game';

export const defaultGameState = new GameState({
	mazeId: '',
	width: 0,
	height: 0,
	gameStatus: GameStatus.ACTIVE,
	charactersPosition: iMap<Role, Point>(),
	blueprint: iList<iList<BlueprintRecord>>()
});

const gameReducer = (state: GameState = defaultGameState, action: GameAction): GameState => {
	switch (action.type) {
		case INIT_STATE: {
			const { state } = action.payload;

			return state;
		}
		case UPDATE_STATE: {
			const { charactersPosition, gameStatus } = action.payload;

			return state.withMutations(state =>
				state.set('charactersPosition', charactersPosition).set('gameStatus', gameStatus)
			) as GameState;
		}
		default:
			return state;
	}
};

export default gameReducer;
