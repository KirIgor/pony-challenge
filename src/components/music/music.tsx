import * as React from 'react';

import './music.css';

interface Props {}

const Music = React.memo(({  }: Props) => {
	return (
		<div>
			<iframe src="music/silence.mp3" allow="autoplay" id="audio" style={{ display: 'none' }} />
			<audio id="music" autoPlay src="music/bundle.mp3" />
		</div>
	);
});

export default Music;
