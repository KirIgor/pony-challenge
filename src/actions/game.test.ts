import { Dispatch } from 'redux';

import { PonyName, Direction, GameState } from '../types/index';
import { StoreState } from '../store/store';
import { defaultGameState } from '../reducers/game';
import { defaultStatisticState } from '../reducers/statistic';
import { defaultMusicState } from '../reducers/music';
import { init, move, GameAction } from './game';

jest.mock('../App');

const getState: () => StoreState = () => ({
	game: defaultGameState,
	statistic: defaultStatisticState,
	music: defaultMusicState
});

describe('game action creators', () => {
	describe('init action creator', () => {
		it('should dispatch correct action', async () => {
			const dispatchResults: any = [];
			const dispatch = jest.fn(x => dispatchResults.push(x));

			await init(15, 15, PonyName.APPLEJACK, 0)(dispatch, getState);

			expect(dispatch.mock.calls.length).toBe(1);
			expect(dispatchResults[0]).toMatchSnapshot();
		});
	});

	describe('move action creator', () => {
		it('should throw error when maze id is empty string', () => {
			expect(async () => {
				await move(Direction.EAST)((() => {}) as Dispatch<GameAction>, getState);
			}).toThrow;
		});

		it('should dispatch correct action', async () => {
			const dispatchResults: any = [];
			const dispatch = jest.fn(x => dispatchResults.push(x));

			const state = {
				game: defaultGameState,
				statistic: defaultStatisticState,
				music: defaultMusicState
			};
			state.game = state.game.set('mazeId', 'test-id') as GameState;
			const getState: () => StoreState = () => state;

			await move(Direction.EAST)(dispatch, getState);

			expect(dispatch.mock.calls.length).toBe(1);
			expect(dispatchResults[0]).toMatchSnapshot();
		});
	});
});
