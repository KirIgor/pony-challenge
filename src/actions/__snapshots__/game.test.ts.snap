// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game action creators init action creator should dispatch correct action 1`] = `
Object {
  "payload": Object {
    "state": Object {
      "blueprint": Immutable.List [
        Immutable.List [
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 6,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 6,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 2,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 6,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 2,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 0,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 4,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 5,
                },
                Immutable.Map {
                        sides: 1,
                },
                Immutable.Map {
                        sides: 3,
                },
        ],
        Immutable.List [
                Immutable.Map {
                        sides: 9,
                },
                Immutable.Map {
                        sides: 9,
                },
                Immutable.Map {
                        sides: 8,
                },
                Immutable.Map {
                        sides: 9,
                },
                Immutable.Map {
                        sides: 12,
                },
                Immutable.Map {
                        sides: 8,
                },
                Immutable.Map {
                        sides: 12,
                },
                Immutable.Map {
                        sides: 8,
                },
                Immutable.Map {
                        sides: 13,
                },
                Immutable.Map {
                        sides: 8,
                },
                Immutable.Map {
                        sides: 9,
                },
                Immutable.Map {
                        sides: 8,
                },
                Immutable.Map {
                        sides: 8,
                },
                Immutable.Map {
                        sides: 9,
                },
                Immutable.Map {
                        sides: 10,
                },
        ],
],
      "charactersPosition": Immutable.Map {
        pony: Object {
                "x": 5,
                "y": 4,
              },
        domokun: Object {
                "x": 10,
                "y": 4,
              },
        exit: Object {
                "x": 8,
                "y": 14,
              },
},
      "height": 15,
      "mazeId": "c6d8aefa-23fb-42a3-8eeb-347ab807de0d",
      "width": 15,
    },
  },
  "type": "INIT_STATE",
}
`;

exports[`game action creators move action creator should dispatch correct action 1`] = `
Object {
  "payload": Object {
    "charactersPosition": Immutable.Map {
      pony: Object {
            "x": 5,
            "y": 4,
          },
      domokun: Object {
            "x": 10,
            "y": 4,
          },
      exit: Object {
            "x": 8,
            "y": 14,
          },
},
  },
  "type": "UPDATE_CHARACTERS_POSITION",
}
`;
